runtime/src/comm/ugni/comm-ugni.c

```diff
@@ -7144,9 +7173,16 @@ void fork_amo(fork_t* p_rf_req, c_nodeid_t locale)
   // Make sure that, if we need a result, it is in memory known to the
   // NIC.
   //
-  if (p_rf_req->a.res != NULL
-      && mreg_for_local_addr(p_rf_req->a.res) == NULL) {
-    CHPL_INTERNAL_ERROR("fork_amo(): result address is not NIC-registered");
+  // Note: the AMO command code 0x1 bit is clear for the 32-bit ones and
+  // set for the 64-bit ones.  (See the fork_amo_cmd_t declaration.)
```

Comment:
This feels quite subtle. Not sure I have a better suggestion, but I worry this is too subtle.

Reply:
The alternative is either a honking big if- or switch-stmt, or a little vector of sizes indexed by the command code, similar to [`nic_amos_ari_fetch[]`](https://github.com/chapel-lang/chapel/blob/0ad48a037f2d5cf6b87ba20e824ec59e5e53b2cc/runtime/src/comm/ugni/comm-ugni.c#L1135). Name your poison (I like the array better).

Actually, while typing this it occurred to me that we could reorder the existing command codes to put all the 32-bit ones first. Then it would be easy to check whether a non-specific code was for a 32-bit operation. But then we'd have to reorder the elements of `nic_amos_ari_fetch[]` and `nic_amos_ari[]` and also, just because I'm that way, any preexisting switch-stmts on command codes. At least the latter sounds painful.

FWIW, the GNI opcodes that correspond to our AMO command codes are defined as having a specific bit that indicates 32- vs. 64-bit. But that's in the private decls, not the public ones.

Reply:
@ronawho Any feedback on these alternatives? Does the "little vector of sizes" idea seem okay?

Reply:
I'm not sure. I _think_ I'd like vector of sizes best, but not positive.

Reply:
And we almost don't even care about getting the size exactly right here (since we're going to round up), so I kinda wonder if we could just pass in 8 bytes and ignore the actual size.

Reply:
Two things: first, I don't think my cache-line round-up logic is correct. It's just rounding up the length. I suspect what it ought to do instead is to round up the transfer's ending address. Second, I don't think it's safe to just check an 8-byte length, because that will break for any 4-byte AMO whose target address is the last 4 bytes in a cache line.

Reply:
Ah, yeah that's a good point. Let's just leave it as-is then.

Reply:
I had an idea for an improvement overnight. I'll define a macro-function, tentatively `IS_32BIT_AMO_CMD(cmd)`, right after the `fork_amo_cmd_t` enum declaration, and then invoke that down in `fork_amo()` where the above `(...cmd & 0x1) == 0` test is now. The actual logic of the check will be the same, but at least the enum decl and the logic will then be right together and there will be less likelihood of messing it up in the future.  (Not that comm=ugni has a lot of "future" left, but still there are better and worse ways to code this sort of thing.)